server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,581 @@
+#! /usr/bin/env python3
+
+import os
+import sys
+import glob
+import tempfile
+
+from pathlib import Path
+from pbench.server import PbenchServerConfig
+from pbench.common.exceptions import BadConfig
+from pbench.common.logger import get_pbench_logger
+from pbench.server.database import init_db
+from pbench.server.report import Report
+from configparser import ConfigParser, NoSectionError, NoOptionError
+from pbench.server.hierarchy import (
+    Hierarchy,
+    ArchiveHierarchy,
+    ControllerHierarchy,
+    IncomingHierarchy,
+    ResultsHierarchy,
+    UserHierarchy,
+)
+
+
+_NAME_ = "pbench-audit-server"
+
+
+class PbenchMDLogConfig(object):
+    """A simple class to wrap a ConfigParser object in order to query a specific
+    metadata.log file for a given tarball.
+    """
+
+    def __init__(self, cfg_name):
+        self.conf = ConfigParser()
+        self.files = self.conf.read([cfg_name])
+
+    def get(self, *args, **kwargs):
+        return self.conf.get(*args, **kwargs)
+
+
+def verify_valid_controllers(hier, controllers):
+    """Find all the non-directory files at the same level of the controller
+    directories and report them, and find all the normal controller directories.
+    """
+    for controller_path in controllers:
+        if controller_path.is_dir():
+            hier.add_controller(controller_path.name)
+        else:
+            hier.add_bad_controller(controller_path)
+    return 0
+
+
+# Archive Hierarchy
+def verify_subdirs(hier, controller, directories):
+    linkdirs = sorted(hier.config.LINKDIRS.split(" "))
+    if directories:
+        for dirent in directories:
+            if all(
+                [
+                    Path(hier.path, controller, dirent).exists(),
+                    dirent != "_QUARANTINED",
+                    not dirent.startswith("WONT-INDEX"),
+                ]
+            ):
+                if dirent not in linkdirs:
+                    hier.add_unexpected_entries(
+                        Hierarchy.UNEXPECTED_DIRS, controller, dirent
+                    )
+    else:
+        hier.add_unexpected_entries(
+            Hierarchy.SUBDIR_STATUS_INDICATORS, controller, "subdirs"
+        )
+
+    return 0
+
+
+def verify_prefixes(hier, controller):
+    prefix_dir = Path(hier.path, controller, ".prefix")
+    if not prefix_dir.exists():
+        return 0
+    if not prefix_dir.is_dir():
+        hier.add_unexpected_entries(
+            Hierarchy.PREFIX_STATUS_INDICATORS, controller, "prefix_dir"
```

Comment:
I don't think the original shell script even cares if there's a `.prefix` that's not a directory; it seems to only care whether `$ARCHIVE/${controller}/.prefix/${tb}.prefix` exists. That's not to say we necessarily _shouldn't_ care, but this function seems to be doing a lot of checking that's not in the original shell script and very little of what _was_ in the original shell script. If you're really intending to be making functional changes, it might be nice to have that explained and isolated from the basic shell -> Python translation.